### ISSUE TITLE
Extend chord space code with functional harmony

### ISSUE THREAD

**Issue by gogins** [gogins/csound-extended#136]
![image](https://user-images.githubusercontent.com/6325029/81000651-9778fc00-8e14-11ea-8481-bce85b4d6ddb.png)

This does not include very modal music, modulations, or extended tonality. It has the considerable virtues of pretty correctly representing what we call "classical music" as well as much folk and popular music, and above all of being based upon the _evidence_ of these musics. This is from Tymoczko. As he also invented the whole idea of chord space and with some others analyzed its fundamental symmetries, it is interesting to note that each move in this graph can be represented by a move in chord space.

Therefore a Markov chain implementing this graph by means of transformations in chord space could be developed as a means of generating, stochastically or by recursion, "well formed" chord progressions.

I don't know how completely I can implement this, but anything along these lines that is consistently well-formed should be _quite_ useful.

**Comment by gogins** [comment]
Playing with this shows a convincing subset of tonal progressions and modulations no matter what order the transformations come in. Seems to come down to simply avoiding the Tn not in the list.

This is great for triadic progressions in the major scale, but there is the minor scale and higher arities of chord as well.

**Comment by gogins** [comment]
I have produced an improved (for my purposes) version of Tymoczko's graph:

![image](https://user-images.githubusercontent.com/6325029/81313028-53c10500-9055-11ea-9ef1-fa61d72e8c36.png)

Green ovals are tonic, blue subdominant, red dominant. Chord progressions are indicated by arrows. Start with a green arrow from I, end with a red arrow to I. In between, arrows may be applied singly or in chains. Usually, begin with one or more descending thirds, end with a descending fifth. Replacing the origin of a green arrow with V in a new key either creates a secondary dominant, or initiates a modulation. Replacing the terminus of a red arrow with I in a new key completes a modulation. Note: these descending thirds are parsimonious voice-leadings near the axis of the orbifold for trichords. Arrows are labeled with corresponding chord transformations in the orbifold; K exchanges the first two voices (taking a major chord in OP form to its relative minor), Tn is transposition by semitones. This is the kernel of tonality. One brutal simplification: transformations use vii for vii°.

**Comment by gogins** [comment]
The thing I've learned here is that randomly choosing transformations from the graph produces much more "harmonious" and even almost "tonal" sounding music that randomly choosing the chords from the graph.

**Comment by gogins** [comment]
First order Markov transition frequencies:
```
     I VI IV II VII  V III
I    0  7 18  14  9 53   1
VI   7  0 15  48  9 20   2
IV  34  0  0  14 14 37   0
II   1  1  0   0 19 77   1
VII 82  1  2   0  0 15   1
V   84  8  5   0  2  0   0  
```

**Comment by gogins** [comment]
As I have a "modality" for the Q transformation, I could put in a "scale" for diatonic transpositions. I think this would be really useful. Chord.eOP().DT(scale, n). The scale would not need to be diatonic, actually.

**Comment by gogins** [comment]
Just in case anyone is wondering, including myself, the point of this work is not to imitate Mozart or the Bee Gees, but to create algorithms that I can then creatively vary.

**Comment by gogins** [comment]
Just to be clear, K and T form a group that generates all major and minor triads from one major or minor triad. And the lambdas discussed here, form a group that generates many many "pseudo tonal" progressions without obvious clunkers and no "backwards" root progressions. That this is so depends totally upon the K operator. If there were 2 such operators that formed a group generating only chords belonging to a given scale, e.g. the diminished triad in the regular triadic major scale chords, that also would work. Otherwise, more context is required, such as modality for Q or scale and scale degree for transposition by scale step.

**Comment by gogins** [comment]
As for _direction_ in chord space, I have formulated the hypothesis that it is the breaking of symmetry caused by melody, which is inherently scalar and tonal. Melody returns, and that creates a graph that goes out and comes back. To begin with the melody doubles itself, and then it accompanies itself as counterpoint, and to be good counterpoint it uses certain voice-leadings, and these become chords, and to go out and come back these chords follow Tymoczko's graph and similar graphs.

I do not have this in my software. There is nothing in the ChordSpace module to break the symmetry.

**_Cadence_**, and cadential formulae.

**Comment by gogins** [comment]
So now I have two questions:

1. How to implement tonality via cadence in ChordSpace?
2. How to add melodic motion and ending to ChordSpace?

**Comment by gogins** [comment]
To make ChordSpace do functional harmony, it needs to operate not on chords (which are ambiguous) but on scale degrees/set-classes. On chords by Roman numeral, in other words. And this can be done. For this to work the chord must carry its scale degree with it and this degree must transform with the chord.

(The thought nags that there is some way to do this purely with melody, which would be more in accord with how we actually think musically. Then we could do something like "harmonize this melody".)

It should be possible to take a scale, take every other from scale degree to arity, and thus produce all chords of that scale in order.

Chord.pitches() -- returns all pitches in ascending order of the chord, i.e. in P but not O.
Chord.chordForDegree(degree, arity, generator = 3) -- Assuming Chord is a scale, returns the chord for the scale degree; arity = 3 is triads, 4 is sevenths, 5 is ninths; generator is scale degrees between chord voices. The returned chord is labeled by scale and degree.

**Comment by gogins** [comment]
The above doesn't work, e.g. because "C major" in OP is "G13". But obtaining a scale by name works fine because the tonic of the scale is in the name of the scale, so you can get OP by name and then invert that chord until the tonic is the first voice. 
```
Chord scale(std::string name);
```
Once you have a scale, you can obtain chords by scale degree and transpose chords by scale degree. The scale always has to be known in order to identify the tonic.
```
Chord chord(Chord scale, int degree, int arity)
Chord transpose(Chord scale, chord, scale_steps);
```
I could bring these functions in as members of a new Scale class derived from Chord.

**Comment by gogins** [comment]
Ambiguities are bothersome but also vital. They permit secondary dominants and modulations. See if these can be implemented using `scale`, `chord`, and voiceleadings.

**Comment by gogins** [comment]
Now that I have scales, chords of scales, and transpositions of chords in scales working, I should consider secondary dominants, modulations, and chromatic alterations? 

I wonder how much of that can be done by inserting a "chord space" transformation in place of "functional harmony" transformation....

I know for a fact that some of this can be done by transforming the scale itself such that the chord leaving belongs to both scales.

**Comment by gogins** [comment]
One of my tasks here sure is to identify other groups such as K/Q and K/T that generate at least some useful subset of harmony.

**Comment by gogins** [comment]
![image](https://user-images.githubusercontent.com/6325029/82159685-0ebf7e80-985e-11ea-9d74-759504e84785.png)

**Comment by gogins** [comment]
Down by scale degree 2 modulo 7, up by scale degree 1 modulo 7

**Comment by gogins** [comment]
This is finally actually working. It will definitely be useful. I now have to think about secondary dominants and modulations. Replacing the origin of a green arrow with V in a new key either creates a secondary dominant, or initiates a modulation. Replacing the terminus of a red arrow with I in a new key completes a modulation. This implies that the arrows have to carry the scale as well as the scale degree. Complex numbers?

**Comment by gogins** [comment]
There are many ways of modulating. Here I indicate those that can be used by a common pivot chord whether or not it is enharmonically different.

1. Common chord as pivot chord (usable).
2. Enharmonic alias as pivot chord (usable).
3. Common tone modulation (same as common chord but with one or more chord tones missing) .
4. Chromatic modulation, a common chord tone is chromatically altered, perhaps in passing.
5. Phrase (brute force) modulation (usable).
6. Sequential modulation (usable by common chord or brute force).
7. Chain modulation.

In short, with the current software common chord, enharmonic, brute force, or sequential modulations may be implemented.

One way to implement this is to provide a Scale method that returns for any Chord a list of other Scales to which it belongs.

**Comment by gogins** [comment]
Secondary dominants in the major scale can be VI, VII, I, II, II or dominant sevenths (or higher functions) of those roots. The algorithm for finding them is:

1. Change the chord type to major or dominant.
2. See if it progresses by 4 scale degrees to a chord in the current Scale.

**Comment by gogins** [comment]
I think that once this is completed, chromatic alterations can be handled quite well by looking at the OP orbifold for close voice-leadings to or from the chords of a scale.

**Comment by gogins** [comment]
The core of this functionality is now completed, so I am closing this issue. It will be necessary to extend this to the other scales in `chordForName`, and to check how non-standard scales work, but they _should_ work.

**Comment by gogins** [comment]
As is well known, 12TET is an idealization based on detuning consonant major and minor triads. As such, 12TET is _intrinsically_ atonal.

I wonder what these consonant triads look like in OP with different systems of just intonation. In particular, I wonder if using some system(s) of just intonation can produce a chord space that has some form of intrinsic tonality.

**Comment by gogins** [comment]
I am re-opening this to make the following enhancements:

- [ ] Bring all scales and modes into static storage, with a separate scalesForNames collection. I.e., all scales will be both scales and chords. Custom scales will still be possible.
- [ ] Provide an option for the `modulate` function to assume the pivot chord is a triad even when it has higher arity. This will increase the number of options for modulation.
- [ ] Provide an option for the `modulate` function to recognize the modes of tonicizations, to better model practice.
- [ ] Provide a move voice function that moves by scale step?
